crates/bevy_animation/src/lib.rs

```diff
@@ -479,174 +485,217 @@ impl AnimationPlayer {
+            removed_targets.insert(entity);
         } else {
-            return true;
+            // Push our own name onto the path and record it.
+            new_target.path.parts.push(name.clone());
+            new_targets.insert(entity, new_target);
+        }
+
+        // Enqueue children.
+        if let Some(kids) = kids {
+            worklist.extend(kids);
         }
     }
+
+    // Finally, update the `AnimationTarget` components.
+    for (entity, animation_target) in new_targets.drain() {
+        commands.entity(entity).insert(animation_target);
+    }
+    for entity in removed_targets.drain() {
+        commands.entity(entity).remove::<AnimationTarget>();
+    }
 }
 
-/// System that will play all animations, using any entity with a [`AnimationPlayer`]
-/// and a [`Handle<AnimationClip>`] as an animation root
-#[allow(clippy::too_many_arguments)]
 pub fn animation_player(
     time: Res<Time>,
-    animations: Res<Assets<AnimationClip>>,
-    children: Query<&Children>,
-    names: Query<&Name>,
-    transforms: Query<&mut Transform>,
-    morphs: Query<&mut MorphWeights>,
-    parents: Query<(Has<AnimationPlayer>, Option<&Parent>)>,
-    mut animation_players: Query<(Entity, Option<&Parent>, &mut AnimationPlayer)>,
+    animation_clips: Res<Assets<AnimationClip>>,
+    mut players: Query<&mut AnimationPlayer>,
 ) {
-    animation_players
-        .par_iter_mut()
-        .for_each(|(root, maybe_parent, mut player)| {
-            update_transitions(&mut player, &time);
-            run_animation_player(
-                root,
-                player,
-                &time,
-                &animations,
-                &names,
-                &transforms,
-                &morphs,
-                maybe_parent,
-                &parents,
-                &children,
-            );
-        });
+    for mut player in players.iter_mut() {
+        update_transitions(&mut player, &time);
+
+        let paused = player.paused;
+        if paused {
+            continue;
+        }
+
+        // Update the main animation.
+        update_animation(&mut player.animation, paused, &time, &animation_clips);
+
+        // Update transition animations.
+        for transition in &mut player.transitions {
+            update_animation(&mut transition.animation, paused, &time, &animation_clips)
+        }
+    }
 }
 
-#[allow(clippy::too_many_arguments)]
-fn run_animation_player(
-    root: Entity,
-    mut player: Mut<AnimationPlayer>,
-    time: &Time,
-    animations: &Assets<AnimationClip>,
-    names: &Query<&Name>,
-    transforms: &Query<&mut Transform>,
-    morphs: &Query<&mut MorphWeights>,
-    maybe_parent: Option<&Parent>,
-    parents: &Query<(Has<AnimationPlayer>, Option<&Parent>)>,
-    children: &Query<&Children>,
+pub fn animate_targets(
+    world: &mut World,
+    mut system_state: Local<
+        Option<
+            SystemState<(
+                Query<&AnimationPlayer, Without<AnimationTarget>>,
```

Comment:
nit: Is that a design requirement that any `AnimationPlayer` is not an `AnimationTarget`? If so, should we have a debug system running that detects misconfigured entities?

Reply:
Yeah, we need `AnimationPlayer` to be a separate read-only query, or else the parallel `EntityMut` closure won't be able to access it via shared memory. So we have to statically separate `AnimationPlayer` from `AnimationTarget`. We could have a debug system, but developers aren't intended to add `AnimationTarget` manually at all; Bevy manages it internally.